doc/schema-migration.md

```diff
@@ -0,0 +1,52 @@
+# Schema Migration or Norms
```

Comment:
Like @jsmassa I am not familiar with the term norm and norms have more precise meaning in mathematics. It would be good to explain what the name conveys in terms of meaning or to just say that it is named like this for consistency with Datomic. To name it like this just to use a different word than migration is not ideal for defining terminology. Why can we not just call it a `schema-migration` for instance?

Reply:
I guess the term comes from [conformity](https://github.com/avescodes/conformity) probably the most widely used migration lib for datomic.

Reply:
I can rename it to schema-migrations though

Reply:
The term "conformity" describes much better what the norms are there for. Otherwise norms can be anything depending on the context.

"Schema migration" doesn't cover everything that can be achieved with the new functionality; it's just one use case.

So maybe there should be two documents: 
1) A document explaining the new functionality with the example of schema migrations
   - can be most of the current document
   - maybe called something like "Conformity using norms"
2) A document about schema migrations which would 
  - talk about the difficulties of changing the schema of a database containing data
  - explain in which case it makes sense to use norms for schema migration
  - refer to document 1) as a guide for using norms

Reply:
I think norms here correspond to normalizing/symmetrizing a space in maths and after reading the conformity docs I get better why it is named that way. I agree that a bit more documentation would be helpful. I think the approach corresponds to a meta-editing framework for Datahike. The idempotence guarantee is necessary so you can invoke the norms time and again during transactions, I suppose, but it is not clear to me yet how good of an idea it is to norm your database like this online (i.e. for each transaction that is coming in), for instance how much overhead this causes. In case we would record all incoming data to the transactor (before it is being acted on), in any append-log such as Kafka, ZeroMQ, hitchhiker-tree or replikativ with CDVCS, we could rewrite the database history starting from the inputs to the transactor (raw data and even for transactions that fail), which is more flexible in general (and best practice as "event sourcing"), but this requires full replay. It would be helpful to me if you can describe how this approach would relate to what we have here and maybe also put it in the docs.

Note also that the invariant library covers similar functionality, but it triggers a full Datalog query itself on the transaction data and db-before/db-after pair to validate that the DB is "normed". It does not do rewrites yet though. So the trigger mechanism for norms could be unified with its approach. I am happy to have a call on this if you are interested @TimoKramer. I am fine with keeping things simple for now, but I think it is at least worth considering the different options before shipping the code to users.

Reply:
I think you misunderstood what this is. I rewrote large parts of the doc. Maybe you can understand it better now?

Reply:
I am ok to postpone this design discussion, since I think I am trying to approach this too generally and the main purpose here is a simple approach for schema migration. 

Being able to replay transactions and fix the schema would nonetheless be a more general approach similar to how XTDB handles this behind Kafka and event sourcing. Replaying a transaction log also can solve the memory issue as transactions are being replayed in sequence. Deriving new attributes from the data in the database is more restricted in comparison because you might have thrown away parts of the information needed in your original schema.